src/toast/toaster.js

```diff
@@ -0,0 +1,159 @@
+  static defaultProps = {
+    placement: PLACEMENT.top,
+  };
+
+  state = {
+    toasts: [],
+  };
+
+  dismissHandlers = {};
+
+  toastId: number = 0;
+
+  onClose = (key: React.Key) => {
+    delete this.dismissHandlers[key];
+    this.setState(({toasts}) => ({
+      toasts: toasts.filter(t => {
+        return !(t.key === key);
+      }),
+    }));
+  };
+
+  getOnCloseHandler = (key: React.Key, onClose: () => void) => {
+    return () => {
+      this.onClose(key);
+      typeof onClose === 'function' && onClose();
+    };
+  };
+
+  getToastProps = (props: string | ToastPropsT): ToastPropsAltT => {
+    const key: React.Key = props.key || `toast-${this.toastId++}`;
+    if (typeof props === 'string') {
+      return {children: props, key};
+    }
+    return {...props, key};
+  };
+
+  renderToast = (toastProps: ToastPropsT): React.Node => {
+    const {onClose, children, key, ...rest} = toastProps;
+    return (
+      <Toast {...rest} key={key} onClose={this.getOnCloseHandler(key, onClose)}>
+        {({dismiss}) => {
+          this.dismissHandlers[key] = dismiss;
+          return children;
+        }}
+      </Toast>
+    );
+  };
+
+  show = (props: string | ToastPropsT): React.Key => {
+    const toastProps = this.getToastProps(props);
+    this.setState(({toasts}) => {
+      // const newList = [...toasts];
+      toasts.unshift(toastProps);
+      return {toasts};
+    });
+    return toastProps.key;
+  };
+
+  update = (key: React.Key, props: string | ToastPropsT): void => {
+    this.setState(({toasts}) => {
+      toasts.forEach((t, index, arr) => {
+        if (t.key === key) {
+          arr[index] = {...t, ...this.getToastProps(props), key};
+        }
+      });
+      return {
+        toasts,
+      };
+    });
+  };
+
+  dismiss = (key: React.Key) => {
+    if (this.dismissHandlers[key]) {
+      this.dismissHandlers[key]();
+    }
+  };
+
+  clearAll = () => {
+    Object.keys(this.dismissHandlers).forEach(key => {
+      this.dismissHandlers[key]();
+    });
+  };
+
+  clear = (key?: React.Key) => {
+    key === undefined ? this.clearAll() : this.dismiss(key);
+  };
+
+  getSharedProps = () => {
+    const {placement} = this.props;
+    return {
+      $placement: placement,
+    };
+  };
+
+  render = () => {
+    const sharedProps = this.getSharedProps();
+    // Only render on the browser (portals aren't supported server-side)
+    if (__BROWSER__) {
+      return ReactDOM.createPortal(
+        <Root {...sharedProps}>
+          {this.props.children}
+          {this.state.toasts.map(this.renderToast)}
+        </Root>,
+        // $FlowFixMe
+        document.body,
+      );
+    }
+    return null;
+  };
+}
+
+export const Toaster = ToasterComponent;
+
+export default {
+  create: function create(props?: ToasterPropsT): React.Node {
+    return <ToasterComponent {...props} ref={toasterRef} />;
+  },
+  show: (props: string | ToastPropsAltT): ?React.Key => {
```

Comment:
Should we add an error message if the dev is trying to call this server-side? (maybe even throw an error?) Seems like that could help prevent some headaches.

Reply:
that sounds like a great idea - I'd go as far as throwing an error, to make sure it is never used that way

Reply:
I have the `console.error('No toaster created yet');` in the `__DEV__` mode. Do you mean adding another warning or error in the render of `Toaster`?